src/Microsoft.Health.Fhir.Liquid.Converter/Processors/JsonToHl7v2Processor.cs

```diff
@@ -0,0 +1,179 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using DotLiquid;
+using Microsoft.Health.Fhir.Liquid.Converter.Exceptions;
+using Microsoft.Health.Fhir.Liquid.Converter.Extensions;
+using Microsoft.Health.Fhir.Liquid.Converter.Models;
+using Microsoft.Health.Fhir.Liquid.Converter.Models.Hl7v2;
+using Microsoft.Health.Fhir.Liquid.Converter.Models.Json;
+using Microsoft.Health.Fhir.Liquid.Converter.Parsers;
+using Newtonsoft.Json.Linq;
+using NJsonSchema;
+
+namespace Microsoft.Health.Fhir.Liquid.Converter.Processors
+{
+    public class JsonToHl7v2Processor : BaseProcessor
+    {
+        private readonly IDataParser _parser = new JsonDataParser();
+
+        private string[] _segmentsWithFieldSeparator = new string[] { "MSH", "BHS", "FHS" };
+
+        public JsonToHl7v2Processor(ProcessorSettings processorSettings)
+            : base(processorSettings)
+        {
+        }
+
+        public override string Convert(string data, string rootTemplate, ITemplateProvider templateProvider, TraceInfo traceInfo = null)
+        {
+            object jsonData;
+
+            jsonData = _parser.Parse(data);
+
+            var result = Convert(jsonData, rootTemplate, templateProvider, traceInfo);
+
+            var hl7Message = GenerateHL7Message(JObject.Parse(result));
+
+            var hl7String = ConvertHl7MessageToString(hl7Message);
+
+            return hl7String;
+        }
+
+        public string Convert(JObject data, string rootTemplate, ITemplateProvider templateProvider, TraceInfo traceInfo = null)
+        {
+            var jsonData = data.ToObject();
+            var result = Convert(jsonData, rootTemplate, templateProvider, traceInfo);
+            var hl7Message = GenerateHL7Message(JObject.Parse(result));
+
+            var hl7String = ConvertHl7MessageToString(hl7Message);
+            return hl7String;
+        }
+
+        public Hl7v2Data GenerateHL7Message(JObject transformations)
+        {
+            var hl7Segments = new List<Hl7v2Segment>();
+
+            // validate that the template has a messageDefinition section
+            if (transformations["messageDefinition"] == null)
+            {
+                throw new RenderException(
+                    FhirConverterErrorCode.TemplateRenderingError,
```

Comment:
nit: would the PropertyNotFound code be more appropriate here?

Reply:
that makes sense. updated